### PR TITLE
fix stream problem in Sampler

### PR DESCRIPTION
**Pull request by esmeetu**
Problem description:
When serving a 34B model(like https://huggingface.co/TheBloke/WizardCoder-Python-34B-V1.0-AWQ) with TP4, the first generation is right, but the second generation will raise RuntimeError. What is weird is that https://huggingface.co/TheBloke/WizardCoder-Python-7B-V1.0-AWQ is fine using TP4 and I don't know why sampler is not working with bigger model.
Moreover, this not only have issues with the AWQ model but also with the GPTQ model.

And i found this error is introduced by PR: #1889 and the reason is that the sampler outputs nan logprops which trigger assert error between the 4 workers's output.
I believe there may be a stream conflict issue when employing tensor parallelism. While I am not skillful at PyTorch streams and made several modifications, I discovered that allowing `self._copy_stream` to wait for the current stream resolves the issue. Another effective fix involves adding a `print` statement in the `from_sampling_metadata` method of ` SamplingTesors`.

In conclusion, this is an experimental fix that requires further discussion. 
I am hopeful for a thorough explanation of this matter since I am uncertain about the effectiveness of this fix.

### CODE REVIEW

**Comment by Yard1** [comment]
Hmm that's quite weird, I hadn't ran into that myself with multiple models with different TP sizes. Let me see if I can reproduce!

**Comment by esmeetu** [comment]
@Yard1 No, I use normal openai server without special config. The gpu is 16Gb T4. There is 8830 gpu blocks left after running 34b model using TP4. API is using v1/chat/completions, while the server's chat template is alpaca.

**Comment by ZiyueHuang** [comment]
Maybe related to https://github.com/vllm-project/vllm/pull/2190?

**Comment by Yard1** [comment]
@esmeetu can you try the PR posted by @ZiyueHuang and see if it helps?

**Comment by esmeetu** [comment]
@Yard1 No, it doesn't.

**Comment by esmeetu** [comment]
@Yard1 Got it! It is the pin_memory issue. Commenting out it is ok.🤗

**Comment by esmeetu** [comment]
If that ok, i will update my PR later.

**Comment by Yard1** [comment]
Pinning memory is what makes this non-blocking - removing that will remove any benefit from this. I haven't had a chance to reproduce yet, hopefully that is something that can be fixed without restoring to full synchronization.

**Comment by ZiyueHuang** [comment]
@esmeetu  Could you please try https://github.com/vllm-project/vllm/pull/2190#issuecomment-1862144866? Sorry for the inconvenience, but I don't have the same hardware resources to reproduce.

**Comment by esmeetu** [comment]
> @esmeetu  Could you please try https://github.com/vllm-project/vllm/pull/2190#issuecomment-1862144866? Sorry for the inconvenience, but I don't have the same hardware resources to reproduce.

@ZiyueHuang Can you give me the code. I didn't know how to implement your idea.

**Comment by esmeetu** [comment]
> Pinning memory is what makes this non-blocking - removing that will remove any benefit from this. I haven't had a chance to reproduce yet, hopefully that is something that can be fixed without restoring to full synchronization.

Oh, I see. thanks for explanation!

**Comment by ZiyueHuang** [comment]
Sure, I'll update the PR.

Maybe the root cause is also due to the cpu pin-memory reuse, but I'm not 100% sure, note that cpu pin_memory tensor is allocated under the `main stream`, while the sampling tensors are allocated under `copy_stream`. I'll take a look on CachingHostAllocator.

**Comment by Yard1** [comment]
@esmeetu Could you try printing out `self._copy_stream.device` in the sampler forward pass and let me know what it is across the ranks?

**Comment by esmeetu** [comment]
@Yard1 
```

INFO 12-19 14:27:21 async_llm_engine.py:379] Received request cmpl-45f72b62419d4d7f8338cd486c33a6c6: prompt: 'Below is an instruction that describes a task. Write a response that appropriately completes the request.\n\n### Instruction:\nping### Response:\n', sampling params: SamplingParams(n=1, best_of=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=0.0, top_p=1.0, top_k=-1, min_p=0.0, use_beam_search=False, length_penalty=1.0, early_stopping=False, stop=[], stop_token_ids=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=128, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True), prompt token ids: [1, 13866, 338, 385, 15278, 393, 16612, 263, 3414, 29889, 14350, 263, 2933, 393, 7128, 2486, 1614, 2167, 278, 2009, 29889, 13, 13, 2277, 29937, 2799, 4080, 29901, 13, 15702, 2277, 29937, 13291, 29901, 13].

INFO 12-19 14:27:21 llm_engine.py:661] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 1 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.2%, CPU KV cache usage: 0.0%

(RayWorkerVllm pid=795641) cuda:1 [repeated 2x across cluster] (Ray deduplicates logs by default. Set RAY_DEDUP_LOGS=0 to disable log deduplication, or see https://docs.ray.io/en/master/ray-observability/ray-logging.html#log-deduplication for more options.)

INFO 12-19 14:27:21 async_llm_engine.py:111] Finished request cmpl-45f72b62419d4d7f8338cd486c33a6c6.

INFO:     30.90.7.99:61106 - "POST /v1/chat/completions HTTP/1.1" 200 OK

INFO 12-19 14:27:36 async_llm_engine.py:379] Received request cmpl-056295b4355146358233e8c6b8b85916: prompt: 'Below is an instruction that describes a task. Write a response that appropriately completes the request.\n\n### Instruction:\nping### Response:\n', sampling params: SamplingParams(n=1, best_of=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=0.0, top_p=1.0, top_k=-1, min_p=0.0, use_beam_search=False, length_penalty=1.0, early_stopping=False, stop=[], stop_token_ids=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=128, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True), prompt token ids: [1, 13866, 338, 385, 15278, 393, 16612, 263, 3414, 29889, 14350, 263, 2933, 393, 7128, 2486, 1614, 2167, 278, 2009, 29889, 13, 13, 2277, 29937, 2799, 4080, 29901, 13, 15702, 2277, 29937, 13291, 29901, 13].

Exception in callback functools.partial(<function _raise_exception_on_finish at 0x7fbda996ee60>, request_tracker=<vllm.engine.async_llm_engine.RequestTracker object at 0x7fbe49ad5870>)

handle: <Handle functools.partial(<function _raise_exception_on_finish at 0x7fbda996ee60>, request_tracker=<vllm.engine.async_llm_engine.RequestTracker object at 0x7fbe49ad5870>)>

Traceback (most recent call last):

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 28, in _raise_exception_on_finish

    task.result()

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 359, in run_engine_loop

    has_requests_in_progress = await self.engine_step()

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 338, in engine_step

    request_outputs = await self.engine.step_async()

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 191, in step_async

    output = await self._run_workers_async(

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 227, in _run_workers_async

    assert output == other_output

AssertionError



The above exception was the direct cause of the following exception:



Traceback (most recent call last):

  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 37, in _raise_exception_on_finish

    raise exc

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 32, in _raise_exception_on_finish

    raise AsyncEngineDeadError(

vllm.engine.async_llm_engine.AsyncEngineDeadError: Task finished unexpectedly. This should never happen! Please open an issue on Github. See stack trace above for the actual cause.

INFO 12-19 14:27:36 async_llm_engine.py:134] Aborted request cmpl-056295b4355146358233e8c6b8b85916.

INFO:     30.90.7.99:61102 - "POST /v1/chat/completions HTTP/1.1" 500 Internal Server Error

ERROR:    Exception in ASGI application

Traceback (most recent call last):

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 28, in _raise_exception_on_finish

    task.result()

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 359, in run_engine_loop

    has_requests_in_progress = await self.engine_step()

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 338, in engine_step

    request_outputs = await self.engine.step_async()

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 191, in step_async

    output = await self._run_workers_async(

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 227, in _run_workers_async

    assert output == other_output

AssertionError



The above exception was the direct cause of the following exception:



Traceback (most recent call last):

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 426, in run_asgi

    result = await app(  # type: ignore[func-returns-value]

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__

    return await self.app(scope, receive, send)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/fastapi/applications.py", line 1106, in __call__

    await super().__call__(scope, receive, send)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/starlette/applications.py", line 122, in __call__

    await self.middleware_stack(scope, receive, send)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__

    raise exc

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__

    await self.app(scope, receive, _send)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/starlette/middleware/cors.py", line 83, in __call__

    await self.app(scope, receive, send)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/aioprometheus/asgi/middleware.py", line 184, in __call__

    await self.asgi_callable(scope, receive, wrapped_send)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__

    raise exc

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__

    await self.app(scope, receive, sender)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__

    raise e

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__

    await self.app(scope, receive, send)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/starlette/routing.py", line 718, in __call__

    await route.handle(scope, receive, send)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/starlette/routing.py", line 276, in handle

    await self.app(scope, receive, send)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/starlette/routing.py", line 66, in app

    response = await func(request)

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/fastapi/routing.py", line 274, in app

    raw_response = await run_endpoint_function(

  File "/home/roy/.conda/envs/ray/lib/python3.10/site-packages/fastapi/routing.py", line 191, in run_endpoint_function

    return await dependant.call(**values)

  File "/home/roy/vllm/vllm/entrypoints/openai/api_server.py", line 438, in create_chat_completion

    return await completion_full_generator()

  File "/home/roy/vllm/vllm/entrypoints/openai/api_server.py", line 384, in completion_full_generator

    async for res in result_generator:

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 445, in generate

    raise e

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 439, in generate

    async for request_output in stream:

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 70, in __anext__

    raise result

  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 37, in _raise_exception_on_finish

    raise exc

  File "/home/roy/vllm/vllm/engine/async_llm_engine.py", line 32, in _raise_exception_on_finish

    raise AsyncEngineDeadError(

vllm.engine.async_llm_engine.AsyncEngineDeadError: Task finished unexpectedly. This should never happen! Please open an issue on Github. See stack trace above for the actual cause.

(RayWorkerVllm pid=795641) cuda:1 [repeated 8x across cluster]

```

**Comment by esmeetu** [comment]
@Yard1 I updated PR, and this modifications can solve my issue. PTAL. I finnaly found that `selected_token_indices` is a tensor, and it will affect `hidden_states`'s computation. So i force that to be same with `hidden_states` device, and works. :)
But also i don't know why this works that i should set `selected_token_indices` to `input_tokens`(same with logits device) device. But when I print `selected_token_indices`'s device before i made modification, they are the same(so weird).
At last, this PR works and looks good although i cannot explain why.

**Comment by Yard1** [comment]
We should definitely keep the `wait_stream`. Let me take a look at the other changes
